go/vt/proto/vtgate/vstreamer/vstreamer.go

```diff
@@ -0,0 +1,389 @@
+}
+
+func (v *VStreamer) initTables(tables []TableConfig) error {
+	if len(v.tables) > 0 {
+		return fmt.Errorf("vstreamer: %d tables already configured", len(v.tables))
+	}
+
+	if len(tables) == 0 {
+		return fmt.Errorf("vstreamer: no tables provided")
+	}
+
+	for _, table := range tables {
+		// basic validation
+		if table.DataType == nil {
+			return fmt.Errorf("vstreamer: table %s.%s has no data type", table.Keyspace, table.Table)
+		}
+
+		if table.Keyspace == "" {
+			return fmt.Errorf("vstreamer: table %v has no keyspace", table)
+		}
+
+		if table.Table == "" {
+			return fmt.Errorf("vstreamer: table %v has no table name", table)
+		}
+
+		// make sure the keyspace and table exist in the cluster
+		// keyspace, ok := v.shardsByKeyspace[table.Keyspace]
+		// if !ok {
+		// 	return fmt.Errorf("vstreamer: keyspace %s not found in the cluster", table.Keyspace)
+		// }
+		//
+		// if !slices.Contains(keyspace, table.Table) {
+		// 	return fmt.Errorf("vstreamer: table %s not found in keyspace %s in the cluster", table.Table, table.Keyspace)
+		// }
+
+		// the key is the keyspace and table name, separated by a period. We use this because the vstream
+		// API uses this as the table name, and it's unique.
+		k := fmt.Sprintf("%s.%s", table.Keyspace, table.Table)
+
+		// if the same table is referenced multiple times in the same stream, only one table will actually
+		// receive events. This prevents users from unknowingly missing events for the second table reference.
+		if _, ok := v.tables[k]; ok {
+			return fmt.Errorf("duplicate table %s in keyspace %s", table.Table, table.Keyspace)
+		}
+
+		// set defaults if not provided
+		if table.Query == "" {
+			table.Query = "select * from " + sqlescape.EscapeID(table.Table)
+		}
+
+		if table.MaxRowsPerFlush == 0 {
+			table.MaxRowsPerFlush = DefaultMaxRowsPerFlush
+		}
+
+		// regardless whether the user provided a pointer to a struct or a struct, we want to store the
+		// underlying type of the struct, so we can create new instances of it later
+		table.underlyingType = reflect.Indirect(reflect.ValueOf(table.DataType)).Type()
+
+		if table.underlyingType.Kind() != reflect.Struct {
+			return fmt.Errorf("vstreamer: data type for table %s.%s must be a struct", table.Keyspace, table.Table)
+		}
+
+		table.shards = make(map[string]shardConfig)
+
+		// initialize the slice containing the batch of rows for this table
+		table.resetBatch()
+
+		// store the table in the map
+		v.tables[k] = &table
+	}
+
+	return nil
+}
+
+func initStateTable(ctx context.Context, session *vtgateconn.VTGateSession, stateKeyspace, stateTable string) error {
+	query := fmt.Sprintf(`create table if not exists %s.%s (
+  keyspace varbinary(512) not null,
+  table varbinary(512) not null,
+  vgtid json,
+  PRIMARY KEY (keyspace, table),
+)`, stateKeyspace, stateTable)
+	_, err := session.Execute(ctx, query, nil)
+	if err != nil {
+		return fmt.Errorf("vstreamer: failed to create state table: %w", err)
+	}
+
+	return nil
+}
```

Comment:
This feels like it _might_ fit in the regular `vreplication` table somehow? Maybe that's shoehorning it where it doesn't go. Some of this decision hinges on whether this is a core vitess thing or a contrib framework.